pkg/kubelet/cm/devicemanager/topology.go

```diff
@@ -0,0 +1,126 @@
+
+	var tempMaskSet []topologymanager.TopologyHint
+	allDeviceSockets := make(map[int]bool)
+
+	firstIteration := true
+	containerRequiresDevice := false
+
+	for resourceObj, amountObj := range container.Resources.Requests {
+		resource := string(resourceObj)
+		amount := int64(amountObj.Value())
+		if m.isDevicePluginResource(resource) {
+			klog.Infof("[devicemanager-topology] %v is a resource managed by device manager.", resource)
+			containerRequiresDevice = true
+			if _, ok := m.healthyDevices[resource]; !ok {
+				klog.Infof("[devicemanager-topology] No healthy devices for resource %v", resource)
+				continue
+			}
+
+			if aligned := checkIfDeviceHasSocketAlignment(resource, m.allDevices); !aligned {
+				klog.Infof("[devicemanager-topology] Device does not have a topology preference")
+				continue
+			}
+			available := m.getAvailableDevices(resource)
+
+			if int64(available.Len()) < amount {
+				klog.Infof("[devicemanager-topology] Requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, amount, available.Len())
+				continue
+			}
+			klog.Infof("[devicemanager-topology] Available devices for resource %v: %v", resource, available)
+
+			deviceSocketAvail := getDevicesPerSocket(resource, available, m.allDevices)
+
+			for socket, amountAvail := range deviceSocketAvail {
+				mask, _ := socketmask.NewSocketMask(int(socket))
+				if amountAvail >= amount {
+					if firstIteration {
+						tempMaskSet = append(tempMaskSet, topologymanager.TopologyHint{SocketAffinity: mask, Preferred: true})
+					} else {
+						isEqual := checkIfMaskEqualsStoreMask(deviceHints, mask)
+						if isEqual {
+							tempMaskSet = append(tempMaskSet, topologymanager.TopologyHint{SocketAffinity: mask, Preferred: true})
+						}
+					}
+				}
+				allDeviceSockets[int(socket)] = true
+			}
+			firstIteration = false
+			deviceHints = tempMaskSet
+			tempMaskSet = nil
+		}
+	}
+
+	if containerRequiresDevice {
+		if len(allDeviceSockets) > 1 {
+			var allDeviceSocketsInt []int
+			for socket := range allDeviceSockets {
+				allDeviceSocketsInt = append(allDeviceSocketsInt, socket)
+			}
+			crossSocketMask, _ := socketmask.NewSocketMask(allDeviceSocketsInt...)
+			deviceHints = append(deviceHints, topologymanager.TopologyHint{SocketAffinity: crossSocketMask, Preferred: false})
+		}
+		klog.Infof("[devicemanager-topology] DeviceHints: %v", deviceHints)
+	}
+
+	return deviceHints, admit
+}
+
+func (m *ManagerImpl) getAvailableDevices(resource string) sets.String {
+	// Gets Devices in use.
+	m.updateAllocatedDevices(m.activePods())
+	devicesInUse := m.allocatedDevices[resource]
+	klog.Infof("Devices in use:%v", devicesInUse)
+
+	// Gets a list of available devices.
+	available := m.healthyDevices[resource].Difference(devicesInUse)
+	return available
+}
+
+func checkIfDeviceHasSocketAlignment(resource string, allDevices map[string][]pluginapi.Device) bool {
+	topologyAligned := false
+	for _, device := range allDevices[resource] {
```

Comment:
the for loop here seems odd.  If we assume that the devices for a given resource all either have or don't have a topology then we should only need to look at the first one.  If they might not have a topology, then shouldn't we return "false" if *any* of them don't have a topology?

Reply:
Changed this and assumed that all or none have topology associated. Thanks!